client/src/templates/Challenges/components/scene/scene.tsx

```diff
@@ -84,142 +86,165 @@ export function Scene({
   ]);
 
   const initBackground = setup.background;
-  const initDialogue = { label: '', text: '', align: 'left' };
-  const initCharacters = setup.characters.map(character => {
-    return {
-      ...character,
-      opacity: character.opacity ?? 1,
-      isTalking: false
-    };
-  });
 
+  // The charactesr are memoized to prevent the useEffect from running on every
+  // render,
+  const initCharacters = useMemo(
+    () =>
+      setup.characters.map(character => {
+        return {
+          ...character,
+          opacity: character.opacity ?? 1,
+          isTalking: false
+        };
+      }),
+    [setup.characters]
+  );
+
+  const [isPlaying, setIsPlaying] = useState(false);
   const [sceneIsReady, setSceneIsReady] = useState(false);
   const [showDialogue, setShowDialogue] = useState(false);
   const [accessibilityOn, setAccessibilityOn] = useState(false);
   const [characters, setCharacters] = useState(initCharacters);
   const [dialogue, setDialogue] = useState(initDialogue);
   const [background, setBackground] = useState(initBackground);
-
-  useEffect(() => {
-    if (isPlaying) {
-      playScene();
-    } else {
-      resetScene();
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [isPlaying]);
+  const startRef = useRef<number>(0);
 
   const audioLoaded = () => {
     setSceneIsReady(true);
   };
 
-  let start = 0;
-  let stopAudio = false;
-
-  // this function exists because we couldn't reliably stop the audio when
-  // playing only part of the audio file. So it would get cut off
-  function maybeStopAudio() {
-    const runningTime = Date.now() - start;
-
-    if (runningTime >= duration) {
-      stopAudio = true;
-      audioRef.current.pause();
-    }
-
-    if (!stopAudio) {
-      window.requestAnimationFrame(maybeStopAudio);
-    }
-  }
-
-  const playScene = () => {
-    setShowDialogue(true);
+  const pause = () => {
+    // Until the play() promise resolves, we can't pause the audio
+    if (canPauseRef.current) audioRef.current.pause();
+    canPauseRef.current = false;
```

Comment:
I am not sure if this has any impact, but do we not want this:

```suggestion
    if (canPauseRef.current) {
      audioRef.current.pause();
      canPauseRef.current = false;
    }
```

Reply:
They're equivalent, so we can use whichever. If `current` is `true`, then both versions set it to `false`. If `current` is `false`, then the unnested version sets it to `false` again, which doesn't matter.

Reply:
I don't mind which version we go with, honestly.